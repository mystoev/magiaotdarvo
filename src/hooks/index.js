export { useCategories } from './use-categories';
export { useToken } from './use-login';
export { useProduct } from './use-product';
export { useProducts } from './use-products';
export { useScrollTop } from './use-scroll-top';
export { useSendEmail } from './use-send-email';
export { useUpdateData, useAddData, useUploadImages } from './use-update-data';
