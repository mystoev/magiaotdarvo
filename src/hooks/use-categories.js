import { useQuery } from 'react-query';
import axios from 'axios';

import { IS_PROD, LOCAL_SERVER } from '../constants';

const CATEGORIES_PHP = 'categories.php';
const CATEGORIES_KEY = 'categories';

const getCategories = async () => {
  const endpoint = IS_PROD ? CATEGORIES_PHP : LOCAL_SERVER + CATEGORIES_PHP;
  const { data } = await axios.get(endpoint);

  return JSON.parse(data);
};

export const useCategories = () => useQuery(CATEGORIES_KEY, () => getCategories());
