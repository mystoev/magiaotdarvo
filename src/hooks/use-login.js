import { useState } from 'react';
import axios from 'axios';

const LOGIN_FILE = 'login.php';

export const login = async (user, password) => {
  try {
    const { data } = await axios.post(LOGIN_FILE, { user, password });
    if (data) {
      localStorage.setItem('token', data);
      return data;
    }
  } catch {
    return null;
  }
};

export const useToken = () => {
  const [token, setToken] = useState(localStorage.getItem('token'));

  return [token, setToken];
};
