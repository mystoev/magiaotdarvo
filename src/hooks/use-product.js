import { useQuery } from 'react-query';
import axios from 'axios';

import { IS_PROD, PRODUCT_PHP, LOCAL_SERVER } from '../constants/data';

const getProduct = async ({ categoryName, productName }) => {
  const endpoint = IS_PROD ? PRODUCT_PHP : LOCAL_SERVER + PRODUCT_PHP;
  const { data } = await axios.get(endpoint + `?category=${categoryName}&product=${productName}`);

  return JSON.parse(data);
};

export const useProduct = ({ categoryName, productName }) =>
  useQuery(`${categoryName}/${productName}`, () => getProduct({ categoryName, productName }));
