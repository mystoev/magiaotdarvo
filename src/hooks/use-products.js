import { useQuery } from 'react-query';
import axios from 'axios';

import { IS_PROD, PRODUCTS_PHP, LOCAL_SERVER } from '../constants/data';

const getProducts = async (categoryName) => {
  const endpoint = IS_PROD ? PRODUCTS_PHP : LOCAL_SERVER + PRODUCTS_PHP;
  const { data } = await axios.get(endpoint + `?category=${categoryName}`);

  return JSON.parse(data);
};

export const useProducts = ({ categoryName }) =>
  useQuery(categoryName, () => getProducts(categoryName));
