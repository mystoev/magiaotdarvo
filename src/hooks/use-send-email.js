import { useMutation } from 'react-query';
import axios from 'axios';

import { EMAIL_FILE } from '../constants/data';
import { selectToken } from '../selectors/token';

const sendEmail = async ({ name, email, phone, message, productLink }) => {
  const token = selectToken();
  const payload = {
    key: token,
    subject: `Запитване от ${name}`,
    message: `
      --------
      Имейл: ${email}
      Телефон: ${phone} ${productLink ? `\n      Относно: ${productLink}` : ''}
      --------
      ${message}
    `
  };

  try {
    const { data } = await axios.post(EMAIL_FILE, payload);
    console.log(data);
  } catch (error) {
    console.log('ERROR SENDING EMAIL');
    console.log(error);
  }
};

export const useSendEmail = (options) => useMutation(sendEmail, options);
