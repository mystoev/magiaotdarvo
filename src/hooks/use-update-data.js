import { UPDATE_FILE, ADD_FILE, UPLOAD_FILE, IS_PROD, LOCAL_SERVER } from '../constants';
import axios from 'axios';
import { selectToken } from '../selectors/token';

export const useUpdateData = async ({
  category,
  productName,
  newProductName,
  description,
  imagesColumn
}) => {
  const token = selectToken();
  const payload = {
    key: token,
    category,
    productName,
    newProductName,
    description,
    imagesColumn
  };

  const endpoint = IS_PROD ? UPDATE_FILE : LOCAL_SERVER + UPDATE_FILE;
  const { data } = await axios.post(endpoint, payload);
  return data;
};

export const useAddData = async ({ category, newProductName, description, images }) => {
  const token = selectToken();
  const payload = {
    key: token,
    category,
    newProductName,
    description,
    images
  };

  const endpoint = IS_PROD ? ADD_FILE : LOCAL_SERVER + ADD_FILE;
  const { data } = await axios.post(endpoint, payload);
  return data;
};

export const useUploadImages = async ({ category, productName, images }) => {
  const token = selectToken();
  const payload = {
    key: token,
    category,
    productName,
    images
  };

  const endpoint = IS_PROD ? UPLOAD_FILE : LOCAL_SERVER + UPLOAD_FILE;
  const { data } = await axios.post(endpoint, payload);
  return data;
};
