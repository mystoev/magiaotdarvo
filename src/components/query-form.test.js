import React from 'react';
import { fireEvent } from '@testing-library/react';
import { renderWithTheme } from '../../test/utils';
import { MemoryRouter as Router, useNavigate } from 'react-router-dom';
import { QueryClient, QueryClientProvider } from 'react-query';

import { useSendEmail } from '../hooks/use-send-email';
import QueryForm from './query-form';

jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-router-dom'),
  useNavigate: jest.fn()
}));
jest.mock('../hooks/use-send-email', () => ({
  useSendEmail: jest.fn().mockReturnValue({ mutate: jest.fn() })
}));

describe('<QueryForm />', () => {
  const queryClient = new QueryClient();
  const renderQueryForm = (slim) =>
    renderWithTheme(
      <QueryClientProvider client={queryClient}>
        <Router>
          <QueryForm slim={slim} />
        </Router>
      </QueryClientProvider>
    );

  const fillForm = (form) => {
    fireEvent.change(form.queryByAltText('name-field'), { target: { value: 'some-name' } });
    fireEvent.change(form.queryByAltText('email-field'), { target: { value: 'email-name' } });
    fireEvent.change(form.queryByAltText('phone-field'), { target: { value: 'phone-name' } });
    fireEvent.change(form.queryByText('', { selector: 'textarea' }), {
      target: { value: 'message-name' }
    });
  };

  it('should render', () => {
    const { container } = renderQueryForm(false);

    expect(container).toMatchSnapshot();
  });

  it('should render as slim', () => {
    const { container } = renderQueryForm(true);

    expect(container).toMatchSnapshot();
  });

  it('should have the button disabled', () => {
    const { queryByRole } = renderQueryForm();

    fireEvent.click(queryByRole('button'));

    expect(queryByRole('button')).toBeDisabled();
  });

  it('when info is typed, should enable button', () => {
    const form = renderQueryForm();

    fillForm(form);

    expect(form.queryByRole('button')).not.toBeDisabled();
  });

  it('should mutate on button click', () => {
    const mockMutate = jest.fn();
    useSendEmail.mockReturnValue({ mutate: mockMutate });
    const form = renderQueryForm();

    fillForm(form);

    fireEvent.click(form.queryByRole('button'));

    expect(mockMutate).toHaveBeenCalled();
  });

  it('should navigate away after email sent', () => {
    const mockNavigate = jest.fn();
    useNavigate.mockReturnValue(mockNavigate);
    useSendEmail.mockImplementation(({ onSuccess }) => {
      onSuccess();
      return { mutate: jest.fn() };
    });
    const form = renderQueryForm();

    fillForm(form);

    fireEvent.click(form.queryByRole('button'));

    expect(mockNavigate).toHaveBeenCalledWith('/email-sent');
  });
});
