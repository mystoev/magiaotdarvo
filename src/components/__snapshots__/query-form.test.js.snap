// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<QueryForm /> should render 1`] = `
.c1 {
  border: 0px;
  border-radius: 3px;
  font-size: 1em;
  font-weight: bold;
  font-family: azbuki;
  background-color: #EEC61F;
  color: #1C1311;
  padding: 15px 40px;
  cursor: pointer;
}

.c1:disabled {
  background-color: grey;
  opacity: 0.5;
  cursor: auto;
}

.c0 {
  margin: 50px 0px;
  display: flex;
  flex-direction: column;
  align-items: center;
  gap: 20px;
}

.c0 p {
  max-width: 800px;
  text-indent: 50px;
  margin: 0px 20px;
}

.c0 input,
.c0 textarea {
  font-family: Verdana;
  font-size: 0.8em;
}

@media only screen and (max-width:768px) {
  .c1 {
    padding: 10px;
    white-space: nowrap;
  }
}

<div>
  <div
    class="c0"
  >
    <p>
      Изпрати запитване, като попълниш формата по-долу, след което ще получиш обаждане за детайли. Или директно се обади на телефон 0890932001.
    </p>
    <fieldset>
      <label>
        Име
      </label>
      <input
        alt="name-field"
        type="text"
        value=""
      />
    </fieldset>
    <fieldset>
      <label>
        Вашият имейл
      </label>
      <input
        alt="email-field"
        type="text"
        value=""
      />
    </fieldset>
    <fieldset>
      <label>
        Телефон за контакт
      </label>
      <input
        alt="phone-field"
        type="text"
        value=""
      />
    </fieldset>
    <fieldset>
      <label>
        Съобщение
      </label>
      <textarea
        alt="message-field"
      />
    </fieldset>
    <button
      class="c1"
      disabled=""
    >
      Изпрати
    </button>
  </div>
</div>
`;

exports[`<QueryForm /> should render as slim 1`] = `
.c1 {
  border: 0px;
  border-radius: 3px;
  font-size: 1em;
  font-weight: bold;
  font-family: azbuki;
  background-color: #EEC61F;
  color: #1C1311;
  padding: 15px 40px;
  cursor: pointer;
}

.c1:disabled {
  background-color: grey;
  opacity: 0.5;
  cursor: auto;
}

.c0 input,
.c0 textarea {
  font-family: Verdana;
  font-size: 0.8em;
}

@media only screen and (max-width:768px) {
  .c1 {
    padding: 10px;
    white-space: nowrap;
  }
}

<div>
  <div
    class="c0"
  >
    <fieldset>
      <label>
        Име
      </label>
      <input
        alt="name-field"
        type="text"
        value=""
      />
    </fieldset>
    <fieldset>
      <label>
        Вашият имейл
      </label>
      <input
        alt="email-field"
        type="text"
        value=""
      />
    </fieldset>
    <fieldset>
      <label>
        Телефон за контакт
      </label>
      <input
        alt="phone-field"
        type="text"
        value=""
      />
    </fieldset>
    <fieldset>
      <label>
        Съобщение
      </label>
      <textarea
        alt="message-field"
      />
    </fieldset>
    <button
      class="c1"
      disabled=""
    >
      Изпрати
    </button>
  </div>
</div>
`;
